File: backend/cartella/src/main/java/cit/edu/cartella/service/ProductService.java
package cit.edu.cartella.service;

import cit.edu.cartella.dto.ProductDTO;
import cit.edu.cartella.entity.Product;
import cit.edu.cartella.entity.Vendor;
import cit.edu.cartella.repository.ProductRepository;
import cit.edu.cartella.repository.VendorRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class ProductService {

    private final ProductRepository productRepository;
    private final VendorRepository vendorRepository;

    @Autowired
    public ProductService(ProductRepository productRepository, VendorRepository vendorRepository) {
        this.productRepository = productRepository;
        this.vendorRepository = vendorRepository;
    }

    // Add a product
    public Product addProduct(Long vendorId, Product productDetails) {
        Vendor vendor = vendorRepository.findById(vendorId)
                .orElseThrow(() -> new RuntimeException("Vendor not found"));

        productDetails.setVendor(vendor);
        return productRepository.save(productDetails);
    }

    // Get all products
    public List<Product> getAllProducts() {
        return productRepository.findAll();
    }

    // Get product by ID
    public Optional<Product> getProductById(Long productId) {
        return productRepository.findById(productId);
    }

    // Get products by vendor
    public List<ProductDTO> getProductsByVendor(Long vendorId) {
        Vendor vendor = vendorRepository.findById(vendorId)
                .orElseThrow(() -> new RuntimeException("Vendor not found"));
        
        return productRepository.findByVendor(vendor).stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    private ProductDTO convertToDTO(Product product) {
        return new ProductDTO(
            product.getProductId(),
            product.getName(),
            product.getDescription(),
            product.getPrice(),
            product.getStockQuantity(),
            product.getCategory(),
            product.getImageUrl(),
            product.getVendor().getBusinessName(),
            product.getCreatedAt(),
            product.getUpdatedAt()
        );
    }

    // Get products by category
    public List<ProductDTO> getProductsByCategory(String category) {
        return productRepository.findByCategoryIgnoreCase(category).stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    // Get products by vendor and category
    public List<ProductDTO> getProductsByVendorAndCategory(Long vendorId, String category) {
        return productRepository.findByVendorVendorIdAndCategoryIgnoreCase(vendorId, category).stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    // Update product
    public Optional<Product> updateProduct(Long productId, Product productDetails) {
        return productRepository.findById(productId).map(product -> {
            product.setName(productDetails.getName());
            product.setDescription(productDetails.getDescription());
            product.setPrice(productDetails.getPrice());
            product.setStockQuantity(productDetails.getStockQuantity());
            product.setCategory(productDetails.getCategory());
            if (productDetails.getImageUrl() != null) {
                product.setImageUrl(productDetails.getImageUrl());
            }
            return productRepository.save(product);
        });
    }

    // Delete product
    public boolean deleteProduct(Long productId) {
        if (productRepository.existsById(productId)) {
            productRepository.deleteById(productId);
            return true;
        }
        return false;
    }
}

File: frontend-web/cartella-frontend/src/api/productService.js
import api from './api';

const productService = {
  // Get all products
  getAllProducts: async () => {
    try {
      const response = await api.get('/products');
      return response.data;
    } catch (error) {
      console.error('Error fetching products:', error);
      throw error;
    }
  },

  // Get product by ID
  getProductById: async (productId) => {
    try {
      const response = await api.get(`/products/${productId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching product:', error);
      throw error;
    }
  },

  // Get products by category
  getProductsByCategory: async (category) => {
    try {
      const response = await api.get(`/products/category/${category}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching products by category:', error);
      throw error;
    }
  },

  // Get products by vendor ID
  getProductsByVendor: async (vendorId) => {
    try {
      const response = await api.get(`/products/vendor/${vendorId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching products by vendor:', error);
      throw error;
    }
  },

  // Get products by vendor ID and category
  getProductsByVendorAndCategory: async (vendorId, category) => {
    try {
      const response = await api.get(`/products/vendor/${vendorId}/category/${category}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching products by vendor and category:', error);
      throw error;
    }
  },

  // Search products
  searchProducts: async (query) => {
    try {
      const response = await api.get(`/products/search?q=${query}`);
      return response.data;
    } catch (error) {
      console.error('Error searching products:', error);
      throw error;
    }
  },

  // Create new product
  createProduct: async (productData) => {
    try {
      const response = await api.post('/products', productData);
      return response.data;
    } catch (error) {
      console.error('Error creating product:', error);
      throw error;
    }
  },

  // Update product
  updateProduct: async (productId, productData) => {
    try {
      const response = await api.put(`/products/${productId}`, productData);
      return response.data;
    } catch (error) {
      console.error('Error updating product:', error);
      throw error;
    }
  },

  // Delete product
  deleteProduct: async (productId) => {
    try {
      const response = await api.delete(`/products/${productId}`);
      return response.data;
    } catch (error) {
      console.error('Error deleting product:', error);
      throw error;
    }
  },

  // Upload product image
  uploadProductImage: async (productId, formData) => {
    try {
      const response = await api.post(`/products/${productId}/upload`, formData, {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      });
      return response.data;
    } catch (error) {
      console.error('Error uploading product image:', error);
      throw error;
    }
  }
};

export default productService;